sat/cli/bootprep/input/image.py

```diff
@@ -479,3 +493,124 @@ def completed_successfully(self):
         return self.image_create_success and self.image_configure_success
 
     begin = begin_image_create
+
+
+class IMSInputImage(BaseInputImage):
+    """An input image that specifies an explicit IMS image or recipe as a starting point."""
+
+    @property
+    def ims_data(self):
+        """dict: the data that defines the base IMS image or recipe"""
+        return self.image_data['ims']
+
+    @property
+    def base_is_recipe(self):
+        """bool: whether the starting point is a recipe in IMS or an image"""
+        # is_recipe is a required property
+        return self.ims_data['is_recipe']
+
+    @property
+    def base_description(self):
+        """str: a human-readable description of the base we are starting from"""
+        for param in ('id', 'name'):
+            if param in self.ims_data:
+                return f'{self.base_resource_type} with {param}={self.ims_data[param]}'
+
+        # This shouldn't happen since schema requires 'id' or 'name'
+        return self.base_resource_type
+
+    @cached_property
+    def ims_base(self):
+        """dict: the data for the base IMS recipe or image to build and/or customize"""
+        resource_type = self.base_resource_type
+        try:
+            matching_base_resources = self.ims_client.get_matching_resources(
+                resource_type, resource_id=self.ims_data.get('id'),
+                name=self.ims_data.get('name'))
+        except APIError as err:
+            raise ImageCreateError(err)
+
+        if len(matching_base_resources) == 0:
+            raise ImageCreateError(f'Found no matches for {self.base_description}')
+        elif len(matching_base_resources) > 1:
+            raise ImageCreateError(f'Found {len(matching_base_resources)} matches for {self.base_description}. '
+                                   f'This {self.base_resource_type} must be specified by "id".')
+
+        return matching_base_resources[0]
```

Comment:
This isn't a problem here, but maybe a note for once we adopt Python 3.10 -- this would be a simple candidate for using the new [structural pattern matching syntax](https://peps.python.org/pep-0636/). This could be replaced perhaps somewhat more compactly with:
```python
match len(matchign_base_resources):
    case 1:
        return matching_base_resources[0]
    case 0:
        raise ImageCreateError(f'Found no matches for {self.base_description}')
    case int(n):
        raise ImageCreateError(f'Found {n} matches for {self.base_description}. '
                               f'This {self.base_resource_type} must be specified by "id".')

```

Reply:
Cool, thanks for the heads up. I wasn't aware of this new feature.